feat: surveys on Max AI

lucasheriques opened the pull request:
## Problem

Closes https://github.com/PostHog/posthog/issues/32818

## Changes

Adds AI survey creation to PostHog's Max tool - users can now create and launch surveys through natural language chat.

### What's New
- **AI Survey Creator**: New `create_survey` MaxTool that converts natural language into configured surveys
- **Chat Interface**: Integrated directly in Surveys page with smart suggestions
- **One-Step Launch**: Create and launch surveys instantly or save as drafts

### Key Features
- Supports all question types (NPS, CSAT, feedback, multiple choice, etc.)
- Context-aware (uses existing surveys to avoid duplication)
- Applies team defaults and survey best practices automatically
- Full validation with detailed error handling

### Testing
Navigate to Surveys page and try:
- "Create an NPS survey for checkout users"
- "Create a feedback survey about our dashboard"  
- "Create and launch a quick satisfaction survey"

### Implementation
- New MaxTool backend in `products/surveys/backend/`
- Pydantic schemas for structured LLM output
- Frontend MaxTool integration
- Comprehensive test coverage

*Part of the Max AI initiative to streamline PostHog workflows through natural language.*

## Did you write or update any docs for this change?

- [x] No docs needed for this change

## How did you test this code?

Unit/integration tests for the max too. @kappa90 i'm not that familiar with evals though, maybe we can pair on it and you can show me around? Are they a good fit for this case?

lucasheriques committed:
feat: surveys on Max

kappa90 commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,196 @@
+class SurveyCreatorTool(MaxTool):
+    name: str = "create_survey"
+    description: str = "Create and optionally launch a survey based on natural language instructions"
+    thinking_message: str = "Creating your survey..."
```
No need to add the three dots "..." as they are added automatically in a later step.

kappa90 commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,196 @@
+    You are helping create surveys for this PostHog team.
+
+    Current context:
+    - Total surveys: {total_surveys_count}
```
Minor: we use [moustache formatting](https://github.com/noahmorrison/chevron) in the other prompts. It basically uses three parentheses instead of one : `{{{total_surveys_count}}}` and you specify "moustache" when doing `ChatPromptTemplate.from_messages`, check any node in `ee/hogai/graph` for an example.

kappa90 reviewed (commented):
Works really well! Will review again when it has tests/it's out of draft 🫶

lucasheriques committed:
Merge branch 'master' into feat/surveys-on-maxai

lucasheriques committed:
wip

github-actions[bot] committed:
Update UI snapshots for `chromium` (7)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

lucasheriques committed:
more schema changes

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

lucasheriques committed:
Merge branch 'master' into feat/surveys-on-maxai

lucasheriques committed:
move to constants

lucasheriques committed:
Merge branch 'master' into feat/surveys-on-maxai

lucasheriques committed:
testing and actually retrieving the user

lucasheriques committed:
Merge branch 'master' into feat/surveys-on-maxai

lucasheriques committed:
mock openAPI

lucasheriques committed:
improve mocks

ioannisj reviewed (commented):
Tried testing this locally but I get an error back from max tool - I'm sure I'm not setup correctly. @lucasheriques if you have time we can pair up so that you can help me set this up properly?

marandaneto commented on frontend/src/scenes/surveys/Surveys.tsx:
```diff
@@ -37,12 +38,67 @@ import { SurveysDisabledBanner, SurveySettings } from './SurveySettings'
+
+    return (
+        <MaxTool
+            name="create_survey"
```
should we also have a tooltip pointing to our survey templates?

lucasheriques replied:
sorry, not sure what you mean? like a tooltip, in the Max chat, with our available templates?

marandaneto replied:
<img width="342" alt="Screenshot 2025-06-23 at 10 51 57" src="https://github.com/user-attachments/assets/29a1607f-c49e-4a85-af96-3f8917ddd768" />

i meant this but with a message like: check out our templates as well

sometimes they dont need to reinvent the wheel

marandaneto commented on frontend/src/scenes/surveys/Surveys.tsx:
```diff
@@ -37,12 +38,67 @@ import { SurveysDisabledBanner, SurveySettings } from './SurveySettings'
+                        status: getSurveyStatus(survey),
+                    })),
+                total_surveys_count: surveys.length,
+                user_id: user?.uuid, // Pass the actual user ID to the backend
```
if the user isnt loaded, will the API error? should we prevent the submission instead?

marandaneto commented on frontend/src/scenes/surveys/Surveys.tsx:
```diff
@@ -37,12 +38,67 @@ import { SurveysDisabledBanner, SurveySettings } from './SurveySettings'
+                    // Optionally navigate to the new survey
+                    router.actions.push(urls.survey(toolOutput.survey_id))
+                } else if (toolOutput?.error) {
+                    // Error case - surveys list should already be current
```
what does it mean 'surveys list should already be current'?

marandaneto commented on posthog/constants.py:
```diff
@@ -319,3 +319,32 @@ class FlagRequestType(StrEnum):
 
 INVITE_DAYS_VALIDITY = 3  # number of days for which team invites are valid
+
+# Sync with frontend/src/scenes/surveys/constants.tsx
```
we hould add a comment to frontend/src/scenes/surveys/constants now to sync with this one so tis both ways

marandaneto commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,240 @@
+            user = None
+            user_id = self.context.get("user_id")
+            if not user_id:
+                return "❌ Failed to create survey: User authentication required", {"error": "authentication_required"}
```
see https://github.com/PostHog/posthog/pull/33111/files#r2154356156
we should not even call the backend if theres no user

lucasheriques replied:
agreed. I still left the validation though, but changed the error and moved it to the beginning of the function to validate if it's present

marandaneto commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,240 @@
+                    "survey_id": str(survey.id),
+                    "survey_name": survey.name,
+                    "launched": result.should_launch,
+                    "questions_count": len(survey.questions) if survey.questions else 0,
```
is a survey valid if theres no questions at all? prob not?

lucasheriques replied:
the serializer does check if the survey has question, but i made that check now explicit in the max tool as well

marandaneto commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,240 @@
+                }
+
+        except Exception as e:
+            return f"❌ Failed to create survey: {str(e)}", {"error": str(e)}
```
we can dog food our error tracking here as well

marandaneto commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,240 @@
+
+            return "\n".join(summaries)
+        except Exception:
+            return "Unable to load existing surveys"
```
same https://github.com/PostHog/posthog/pull/33111/files#r2154369070

lucasheriques replied:
agree, done

marandaneto commented on products/surveys/backend/prompts.py:
```diff
@@ -0,0 +1,73 @@
+
+## Current Context
+Team survey configuration: {team_survey_config}
+Existing surveys: {existing_surveys}
```
why do we need existing surveys?

lucasheriques committed:
Merge branch 'master' into feat/surveys-on-maxai

lucasheriques committed:
add verb

lucasheriques commented on frontend/src/scenes/surveys/Surveys.tsx:
```diff
@@ -37,12 +38,67 @@ import { SurveysDisabledBanner, SurveySettings } from './SurveySettings'
+        <MaxTool
+            name="create_survey"
+            displayName="AI Survey Creator"
+            initialMaxPrompt="Create a survey to "
```
agree, added

lucasheriques commented on products/surveys/backend/prompts.py:
```diff
@@ -0,0 +1,73 @@
+Existing surveys: {existing_surveys}
+
+## Output Requirements
+Always respond with valid JSON containing:
```
actually, we're using structured responses, [where we define ahead of time the output we want](https://github.com/PostHog/posthog/blob/feat/surveys-on-maxai/products/surveys/backend/max_tools.py#L57). so no need to add more details here

lucasheriques committed:
dont load max if user is not there

lucasheriques commented on frontend/src/scenes/surveys/Surveys.tsx:
```diff
@@ -37,12 +38,67 @@ import { SurveysDisabledBanner, SurveySettings } from './SurveySettings'
+            }) => {
+                // Handle survey creation result
+                if (toolOutput?.survey_id) {
+                    // Refresh surveys list to show new survey
+                    loadSurveys()
+                    // Optionally navigate to the new survey
+                    router.actions.push(urls.survey(toolOutput.survey_id))
```
actually, we need to do both, though the comments are misleading.

we should still call `loadSurveys` to make sure the list is updated on the new survey. but then, let's redirect to the new survey still. this way, if our customer comes back, the survey is already there (otherwise it won't be).

i'll remove the comments

lucasheriques committed:
update redirection

lucasheriques committed:
sync comment

lucasheriques committed:
validate user_id and move to beginning

lucasheriques committed:
make question check explicit

lucasheriques committed:
use error tracking

lucasheriques committed:
Merge branch 'master' into feat/surveys-on-maxai

lucasheriques committed:
update test

lucasheriques committed:
fix test again

kappa90 commented on frontend/src/scenes/surveys/Surveys.tsx:
```diff
@@ -37,12 +39,72 @@ import { SurveysDisabledBanner, SurveySettings } from './SurveySettings'
+                'Create a quick satisfaction survey for support interactions',
+            ]}
+            context={{
+                existing_surveys: surveys
```
These context parameters (except for `user_id`) are unused in the backend.

kappa90 commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,240 @@
+    You are helping create surveys for this PostHog team.
+
+    Current context:
+    - Total surveys: {{{total_surveys_count}}}
```
+ 1 to this.

lucasheriques replied:
removed, no need for now

kappa90 commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,247 @@
+                return "❌ Failed to create survey: Invalid user", {"error": "invalid_user"}
+
+            # Get team for context
+            team = Team.objects.get(id=self._team_id)
```
Same as above, from that PR you should be able to call `self._team`

lucasheriques replied:
done

kappa90 commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,247 @@
+            survey_data = self._convert_to_posthog_format(result, team)
+
+            # Use the proper serializer for validation and creation
+            from posthog.api.survey import SurveySerializerCreateUpdateOnly
```
Add these imports on top?

kappa90 commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,247 @@
+    def _get_existing_surveys_summary(self) -> str:
+        """Get summary of existing surveys for context."""
+        try:
+            from posthog.models import Survey
```
can we import at the top?

kappa90 commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,247 @@
+            "default_settings": {"type": "popover", "enable_partial_responses": True},
+        }
+
+    def _get_existing_surveys_summary(self) -> str:
```
You already have `existing_surveys` coming from the frontend, use one or the other?

kappa90 reviewed (changes requested):
Works well!

A few comments:

`MaxTool` wraps the whole survey UI, which, when the sidebar is open, shows a red dotted line around the whole view (see below). I think we want to remove that line, but for now, I'd probably just wrap the "New survey" button.

<img width="717" alt="image" src="https://github.com/user-attachments/assets/81850e13-3a62-4f95-a10b-17874c252950" />

Also added a bunch of code comments.

Let me know when it's re-ready and I'll re-review it. Thanks!

lucasheriques committed:
Merge branch 'master' into feat/surveys-on-maxai

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

lucasheriques committed:
review changes

lucasheriques committed:
improve prompt and validation

lucasheriques commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,247 @@
+        """
+        try:
+            user = None
+            user_id = self.context.get("user_id")
```
using it now

lucasheriques committed:
Merge branch 'master' into feat/surveys-on-maxai

lucasheriques committed:
update tests

lucasheriques committed:
Merge branch 'master' into feat/surveys-on-maxai

lucasheriques committed:
Merge branch 'master' into feat/surveys-on-maxai

lucasheriques committed:
evals for surveys

lucasheriques committed:
forgot the file

lucasheriques committed:
those are not needed

lucasheriques committed:
Merge branch 'master' into feat/surveys-on-maxai

lucasheriques committed:
typecheck

marandaneto commented:
@lucasheriques what is missing here?

lucasheriques commented:
@marandaneto it's missing the last review in the evals part I did after Em guided me through it. I [asked the max ai team here](https://posthog.slack.com/archives/C06NZEZ7V3Q/p1752158817124389), will bump again

lucasheriques committed:
Merge branch 'master' into feat/surveys-on-maxai

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

denakorita commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,225 @@
+                return "❌ Failed to create survey: Invalid user", {"error": "invalid_user"}
+
+            # Get team for context
+            team = self._team
```
I am not certain if this is case actually is ever met, afaik there must always be a team.

denakorita replied:
Same for `user_id`

lucasheriques replied:
removed them!

denakorita commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,225 @@
+
+            result = self._create_survey_from_instructions(instructions)
+            try:
+                validated_output = SurveyCreationOutput.model_validate(result.model_dump())
```
Here I think you dont really need to do `model_dump()` and then `model_validate()` again, as far as I see from the `_create_survey_from_instructions` should always return you a valid `SurveyCreationOutput` object, unless the LLM failed. Maybe I am missing something..

lucasheriques replied:
ohhh makes sense, so this roundtrip was unnecessary. just removed and tested, looking good

denakorita commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,225 @@
+        # Set up the LLM with structured output
+        model = (
+            ChatOpenAI(model="gpt-4.1-mini", temperature=0.2)
+            .with_structured_output(SurveyCreationOutput, include_raw=False)
```
This would do the trick for you to get the correct `SurveyCreationOutput` model. So your return type on this func should be `SurveyCreationOutput`

denakorita commented on ee/hogai/eval/eval_surveys.py:
```diff
@@ -0,0 +1,206 @@
+        if issues:
+            return Score(
+                name=self._name(),
+                score=0,
```
How about instead of giving 0, you can calculate the score as the `number_of_matches/total_checks`? So as to reflect that the generation did succeed for some.

lucasheriques replied:
agreed, changed, thanks for the tip!

denakorita reviewed (commented):
Added a few comments

lucasheriques committed:
Merge branch 'master' into feat/surveys-on-maxai

lucasheriques committed:
checks are not needed

lucasheriques committed:
structured output guarantees the correct validation

lucasheriques committed:
change to async instead

lucasheriques commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,225 @@
+
+        return result
+
+    def _run_impl(self, instructions: str) -> tuple[str, dict[str, Any]]:
```
i just changed to async, can you take another look?

github-actions[bot] committed:
Update query snapshots

lucasheriques committed:
add LLM as judge eval

lucasheriques committed:
Merge branch 'master' into feat/surveys-on-maxai

Twixes committed:
Fix non-null `SurveyDisplayConditions.url`

Twixes committed:
Clarify "Appearance" section of surveys settings

This confused me: when I clicked the settings icon on an already-existing survey, I saw "Appearance" – but from that heading I thought this applies to my current survey. Turns out it's _default appearance_. Clarifying here.

Twixes committed:
User verb for tool display name

Twixes committed:
Also rename tool class to verb

Twixes commented on products/surveys/backend/survey_schema.py:
The sneakiest thing of the PR I think is keeping track of the survey definition schema over time. Thios new survey_schema.py looks like it'll be TOUGH to keep in sync. Have you considered making `frontend/src/queries/schema/` the source of truth? Would be awesome to move all the Surveys-specific definitions into a new `frontend/src/queries/schema/schema-surveys.ts` – that way it'll immediately be possible to import the same exact definition in both the frontend and backend (including Max backend)

lucasheriques replied:
ohhh I didn't know I could do that! Yes definitely will do

denakorita commented on ee/hogai/eval/eval_surveys.py:
```diff
@@ -42,6 +43,173 @@ async def call_max_tool(instructions: str) -> dict:
+- poor: The survey is barely relevant to the user's request
+- irrelevant: The survey does not address the user's request at all
+""".strip(),
+            choice_scores={
```
Same for the other eval

denakorita commented on ee/hogai/eval/eval_surveys.py:
```diff
@@ -202,5 +381,11 @@ async def eval_surveys(call_surveys_max_tool):
+            # Test case 4: Comprehensive survey should still be kept short for in-app use
+            EvalCase(
+                input="Create a comprehensive survey to understand user demographics, usage patterns, satisfaction levels, feature preferences, and improvement suggestions",
+                expected={"min_questions": 1},
```
https://www.braintrust.dev/app/PostHog/p/max-ai-surveys/experiments/feat%2Fsurveys-on-maxai-1752602966?c=feat/surveys-on-maxai-1752602543&tg=false&r=d04138cc-eb18-4249-bc44-4166f9267ca6&s=21077eba-50f1-41fa-9e34-7dc3fddf69ae

lucasheriques replied:
fixed!

denakorita reviewed (dismissed):
Great work, just a few comments

Twixes commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,214 @@
+            capture_exception(e, {"team_id": self._team.id})
+            return "Unable to load existing surveys"
+
+    def _convert_to_posthog_format(self, llm_output: SurveyCreationOutput, team: Team) -> dict[str, Any]:
+        """Convert LLM output to PostHog survey format."""
```
What would the difference be if the LLM's output was structured in PostHog's format out of the gate?

Twixes reviewed (dismissed):
It's a great start! Love it, genuinely useful. Just before merging, I think it's worth addressing all the outstanding comments, including the ones from Greptile.

The next steps will be about improving the proposed display conditions I think, as currently they're very low-context/off. But we're just working in the Max AI team to give you better tools to use the real and relevant events and properties, so let's leave that for v2 of this tool.

lucasheriques committed:
refactor tests and schema

lucasheriques committed:
review changes for evalS

lucasheriques commented on ee/hogai/eval/eval_surveys.py:
```diff
@@ -130,45 +297,55 @@ def _run_eval_sync(self, output, expected=None, **kwargs):
+        # Check 1: Has name
+        has_name = bool(survey_output.name)
+        checks.append("name")
+        if has_name:
```
removed all of them

lucasheriques committed:
always set

lucasheriques commented on frontend/src/scenes/surveys/Surveys.tsx:
```diff
@@ -42,7 +44,76 @@ export const scene: SceneExport = {
+            context={{
+                user_id: user.uuid,
+            }}
+            callback={(toolOutput: {
+                survey_id?: string
+                survey_name?: string
+                launched?: boolean
+                error?: string
+            }) => {
+                if (toolOutput?.error || !toolOutput?.survey_id) {
+                    posthog.captureException('survey-creation-failed', {
+                        error: toolOutput.error,
+                    })
+                    return
+                }
+
+                // Refresh surveys list to show new survey, then redirect to it
+                loadSurveys()
+                router.actions.push(urls.survey(toolOutput.survey_id))
+            }}
```
FE is mostly ai-generated! but I was actually using those parameters, but at some point I decided to simplify it

lucasheriques replied:
`error` is being sent on Exceptions in the BE, unless I'm missing something, [it's what this does, right?](https://github.com/PostHog/posthog/blob/feat/surveys-on-maxai/products/surveys/backend/max_tools.py#L96-L104)

now I removed the other unused fields too! thanks for pointing them out

lucasheriques commented on ee/hogai/eval/eval_surveys.py:
```diff
@@ -0,0 +1,206 @@
+                },
+            )
+
+
```
I changed this eval a little bit, and I'm passing the parameter now as expected

lucasheriques committed:
more review changes

lucasheriques committed:
Merge branch 'master' into feat/surveys-on-maxai

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

lucasheriques committed:
review changes

lucasheriques committed:
fix tests

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

lucasheriques committed:
update prompmt

lucasheriques committed:
simplify schema/conversion

lucasheriques commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,222 @@
+            capture_exception(e, {"team_id": self._team.id})
+            return "Unable to load existing surveys"
+
+    def _convert_to_posthog_format(self, llm_output: SurveyCreationSchema, team: Team) -> dict[str, Any]:
```
just did that, agree there's no need. can you take another look?

lucasheriques committed:
Merge branch 'master' into feat/surveys-on-maxai

github-actions[bot] committed:
Update UI snapshots for `chromium` (14)

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (8)

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

github-actions[bot] committed:
Update UI snapshots for `chromium` (14)

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

github-actions[bot] committed:
Update UI snapshots for `chromium` (12)

kappa90 commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,173 @@
+                    survey_data["start_date"] = datetime.now()
+
+                # Create the survey directly using Django ORM
+                survey = await sync_to_async(Survey.objects.create)(team=team, created_by=user, **survey_data)
```
you can use `Survey.objects.acreate` instead of using `sync_to_async`

kappa90 commented on products/surveys/backend/max_tools.py:
```diff
@@ -0,0 +1,173 @@
+        """Get summary of existing surveys for context."""
+        try:
+            # Use sync_to_async to convert the Django QuerySet to async
+            surveys = await sync_to_async(list)(
```
You can do this:

```
surveys = [survey async for Survey.objects.filter(team_id=self._team.id, archived=False)[:5]]
```

kappa90 commented on products/surveys/backend/prompts.py:
```diff
@@ -0,0 +1,116 @@
+**Existing Surveys**:
+{{{existing_surveys}}}
+- Avoid creating duplicate surveys with similar purposes
+- Reference existing survey names for consistency
```
Not sure I understand why this is needed, if we're creating a new survey, why are we referencing old ones?

lucasheriques replied:
right now, we can't really combine two surveys data. So we don't want to encourage people creating say 2 NPS or 2 CSAT surveys, and instead, combine all of them into one survey.

since, this way, it makes it easy to analyze, as we don't have to do cross-surveys data

kappa90 reviewed (approved):
Left a couple comments on async ORM calls, approving to unblock you, good job!

lucasheriques commented on products/surveys/backend/survey_schema.py:
```diff
@@ -0,0 +1,144 @@
+        default=DEFAULT_SURVEY_APPEARANCE["position"]
+    )
+    thankYouMessageHeader: Optional[str] = Field(default=DEFAULT_SURVEY_APPEARANCE["thankYouMessageHeader"])
+    thankYouMessageDescription: Optional[str] = Field(default="We appreciate your feedback.")
```
this is no longer used

lucasheriques commented on products/surveys/backend/prompts.py:
```diff
@@ -0,0 +1,73 @@
+
+## Current Context
+Team survey configuration: {team_survey_config}
+Existing surveys: {existing_surveys}
```
prompt is different now

lucasheriques committed:
schema changes and fixing all tests
